src/lib.rs

```diff
@@ -77,44 +73,153 @@ mod builder_attr;
-                    let builder_creation = struct_info.builder_creation_impl()?;
-                    let conversion_helper = struct_info.conversion_helper_impl()?;
-                    let fields = struct_info.fields.iter().map(|f| struct_info.field_impl(f).unwrap());
-                    let fields = quote!(#(#fields)*);
-                    let build_method = struct_info.build_method_impl();
+        syn::Data::Struct(data) => match &data.fields {
+            syn::Fields::Named(fields) => {
+                let struct_info = struct_info::StructInfo::new(&ast, fields.named.iter())?;
+                let builder_creation = struct_info.builder_creation_impl()?;
+                let conversion_helper = struct_info.conversion_helper_impl()?;
+                let fields = struct_info
+                    .included_fields()
+                    .map(|f| struct_info.field_impl(f).unwrap());
+                let fields = quote!(#(#fields)*);
+                let build_method = struct_info.build_method_impl();
 
-                    quote!{
-                        #builder_creation
-                        #conversion_helper
-                        #( #fields )*
-                        #build_method
-                    }
+                quote! {
+                    #builder_creation
+                    #conversion_helper
+                    #( #fields )*
+                    #build_method
                 }
-                syn::Fields::Unnamed(_) => return Err(Error::new(ast.span(), "SmartBuilder is not supported for tuple structs")),
-                syn::Fields::Unit => return Err(Error::new(ast.span(), "SmartBuilder is not supported for unit structs")),
             }
+            syn::Fields::Unnamed(_) => {
+                return Err(Error::new(
+                    ast.span(),
+                    "TypedBuilder is not supported for tuple structs",
+                ))
+            }
+            syn::Fields::Unit => {
+                return Err(Error::new(
+                    ast.span(),
+                    "TypedBuilder is not supported for unit structs",
+                ))
+            }
+        },
+        syn::Data::Enum(_) => {
+            return Err(Error::new(
+                ast.span(),
+                "TypedBuilder is not supported for enums",
+            ))
+        }
+        syn::Data::Union(_) => {
+            return Err(Error::new(
+                ast.span(),
+                "TypedBuilder is not supported for unions",
+            ))
         }
-        syn::Data::Enum(_) => return Err(Error::new(ast.span(), "SmartBuilder is not supported for enums")),
-        syn::Data::Union(_) => return Err(Error::new(ast.span(), "SmartBuilder is not supported for unions")),
     };
     Ok(data)
 }
+
+// It’d be nice for the compilation tests to live in tests/ with the rest, but short of pulling in
+// some other test runner for that purpose (e.g. compiletest_rs), rustdoc compile_fail in this
```

Comment:
I actually wouldn't mind adding `compiletest_rs`. I was unaware of its existence (ashamed to say I didn't bother to search for something like this), but for this kind of crate it can be very useful, and allow writing tests for many currently manually-tested (==untested) scenarios.

Reply:
I agree, adding compiletest_rs seems like a good idea. I don’t plan to block this on it, though.